sched/addrenv/addrenv.c

```diff
@@ -0,0 +1,523 @@
+ *
+ * Input Parameters:
+ *   ptcb - The tcb of the parent process
+ *   tcb  - The tcb of the child process
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int addrenv_join(FAR struct tcb_s *ptcb, FAR struct tcb_s *tcb)
+{
+  int ret;
+
+  ret = up_addrenv_attach(ptcb, tcb);
+  if (ret < 0)
+    {
+      berr("ERROR: up_addrenv_attach failed: %d\n", ret);
+      return ret;
+    }
+
+  /* Take a reference to the address environment */
+
+  addrenv_take(ptcb->mm_own);
+
+  /* Share the parent's address environment */
+
+  tcb->mm_own = ptcb->mm_own;
+  tcb->mm_curr = tcb->mm_own;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: addrenv_leave
+ *
+ * Description:
+ *   Leave a process's address environment.
+ *
+ * Input Parameters:
+ *   tcb  - The tcb of the process
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int addrenv_leave(FAR struct tcb_s *tcb)
+{
+  int ret;
+
+  /* Detach from the address environment */
+
+  ret = up_addrenv_detach(tcb);
+
+  /* Then drop the address environment */
+
+  addrenv_drop(tcb->mm_own, false);
+  tcb->mm_own = NULL;
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: addrenv_select
+ *
+ * Description:
+ *   Temporarily select a different address environment for the currently
+ *   running process.
+ *
+ * Input Parameters:
+ *   addrenv - The address environment.
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int addrenv_select(FAR struct addrenv_s *addrenv)
+{
+  FAR struct tcb_s *tcb = this_task();
+  addrenv_take(addrenv);
+  tcb->mm_curr = addrenv;
+  return addrenv_switch(tcb);
+}
+
+/****************************************************************************
+ * Name: addrenv_restore
+ *
+ * Description:
+ *   Switch back to the procces's own address environment.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int addrenv_restore(void)
+{
+  FAR struct tcb_s *tcb = this_task();
+  addrenv_give(tcb->mm_curr);
+  tcb->mm_curr = tcb->mm_own;
+  return addrenv_switch(tcb);
```

Comment:
This logic is still a bit faulty. The binfmt (or elf/nxflat) temporary addrenv is freed here and it obviously fails because it is not heap.

Reply:
Fixed by grabbing a reference in nxflat_addrenv_alloc / elf_addrenv_alloc.